[chatgpt] Add support for LocalAI and other compatible services

kaikreuzer opened the pull request:
See https://github.com/openhab/openhab-addons/pull/14809#issuecomment-1665515466.

This PR makes the urls for the inference API and the model options configurable, so that the OpenAI service can be replaced by other API-compatible services, such as LocalAI.

Note: Not yet tested by myself - testing needs to be done before merging!

kaikreuzer commented on bundles/org.openhab.binding.chatgpt/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -19,6 +19,26 @@
+				<default>https://api.openai.com/v1/chat/completions</default>
+				<advanced>true</advanced>
+				<options>
+					<option value="https://api.openai.com/v1/chat/completions">https://api.openai.com/v1/chat/completions</option>
```
@mudler What would be the suggested option here for LocalAI?

mudler replied:
For LocalAI would be enough to point it to the LocalAI service, so for instance, if you are running it locally on the `8080` port would be:

http://localhost:8080/v1/chat/completions

kaikreuzer replied:
Port 8080 is also used by openHAB, so a user running both on the same machine would need to use a non-default port for LocalAI... Sounds as if it is hence difficult to provide a suggestion here.

kaikreuzer replied:
@mudler Could you define an alternative port for LocalAI users to configure, if they are installing it on the same machine as openHAB ( and integrating with it)? If so, I would put that port as a value in here. Using `8080` is simply not working...

kaikreuzer replied:
@mudler I actually failed setting up LocalAI on a different port. I changed the `ADDRESS` parameter in the `.env` file, but it didn't have any effect...

mudler replied:
apologize, I missed this until now - did you tried with docker only?

```
docker run -p 9090:8080 -v $PWD/models:/models -ti --rm quay.io/go-skynet/local-ai:latest --models-path /models --context-size 700 --threads 4
```

kaikreuzer commented on bundles/org.openhab.binding.chatgpt/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -19,6 +19,26 @@
+				<default>https://api.openai.com/v1/models</default>
+				<advanced>true</advanced>
+				<options>
+					<option value="https://api.openai.com/v1/models">https://api.openai.com/v1/models</option>
```
@mudler Is such a url also available from LocalAI, so that the available models can be retrieved? If so, what's the correct url?

mudler replied:
yep it does exist, it would be equivalent to:

http://localhost:8080/v1/models

mudler replied:
the text generation endpoints are documented here : https://localai.io/features/text-generation/ - however there is 1:1 compatibility with OpenAI, so besides the URL and the model there is nothing else to touch! :)

kaikreuzer replied:
Ok, great.
The tricky part might be https://github.com/openhab/openhab-addons/blob/3cd8c3cf7872c30f9b60d72793383758ea055b08/bundles/org.openhab.binding.chatgpt/src/main/resources/OH-INF/thing/thing-types.xml#L55 - we are currently setting the model parameter to `gpt-3.5-turbo` as a default, which won't exist for a LocalAI installation. If users do not adapt this parameter: How will LocalAI react? Will it result in an error when this model is requested or will it silently fall back to another default model?

mudler replied:
by default if no model is specified in the request, or if there is only one installed it will redirect all the requests to it - however in LocalAI the user is free to call models as they like, so they can mimick gpt-4 behind the scene

mudler reviewed (approved):
That's great! thanks @kaikreuzer !

lolodomo commented:
> Not yet tested by myself - testing needs to be done before merging!

Was it finally tested by someone ?
If not, I suggest to tag the PR with the appropriate tag (tests needed).
Is it ready for a review ?

lolodomo commented on bundles/org.openhab.binding.chatgpt/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -19,6 +19,26 @@
+				<limitToOptions>false</limitToOptions>
+			</parameter>
+			<parameter name="modelUrl" type="text" required="false">
+				<label>Base URL</label>
```
Maybe rather "Model URL" ?

lolodomo commented on bundles/org.openhab.binding.chatgpt/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -19,6 +19,26 @@
 				<label>API Key</label>
 				<description>API key to access the account</description>
 			</parameter>
+			<parameter name="apiUrl" type="text" required="false">
```
Please add the new entries in chatgpt.properties file (just regenerate it) and add the 2 new configuration parameters in README.

kaikreuzer commented:
Thanks for your review, @lolodomo! I have addressed all your comments.

> Is it ready for a review ?

I was still waiting for an answer from @mudler [here](https://github.com/openhab/openhab-addons/pull/15385#discussion_r1320853424), but this is a minor thing that imho should not prevent us from merging this PR.

lolodomo commented on bundles/org.openhab.binding.chatgpt/src/main/resources/OH-INF/i18n/chatgpt.properties:
```diff
@@ -12,6 +12,12 @@ thing-type.chatgpt.account.description = Account at OpenAI that is used for acce
 thing-type.config.chatgpt.account.apiKey.description = API key to access the account
+thing-type.config.chatgpt.account.apiUrl.label = API URL
+thing-type.config.chatgpt.account.apiUrl.description = The server API where to reach the AI service.
+thing-type.config.chatgpt.account.apiUrl.option.https\://api.openai.com/v1/chat/completions = https://api.openai.com/v1/chat/completions
```
I have a big doubt. Do you think this entry is valid ?

lolodomo replied:
Maybe you should just omit the entry, translating a URL has no sense by the way.

Same for model URL.

kaikreuzer replied:
Well, this file is being generated, so even if I manually remove it here, it will reappear on the next generation.
I would assume that the entry is correct - if not, we should rather adapt the i18n generation.

lolodomo replied:
The i18n generator is an helper but we have several exceptions. I believe we have to avoid adding translation work in Crowdin when it is useless.

kaikreuzer replied:
Ok, I have removed the two lines.

kaikreuzer commented on bundles/org.openhab.binding.chatgpt/src/main/resources/OH-INF/i18n/chatgpt.properties:
```diff
@@ -12,6 +12,12 @@ thing-type.chatgpt.account.description = Account at OpenAI that is used for acce
 thing-type.config.chatgpt.account.apiKey.description = API key to access the account
+thing-type.config.chatgpt.account.apiUrl.label = API URL
+thing-type.config.chatgpt.account.apiUrl.description = The server API where to reach the AI service.
+thing-type.config.chatgpt.account.apiUrl.option.https\://api.openai.com/v1/chat/completions = https://api.openai.com/v1/chat/completions
```
```suggestion
```

kaikreuzer commented on bundles/org.openhab.binding.chatgpt/src/main/resources/OH-INF/i18n/chatgpt.properties:
```diff
@@ -12,6 +12,12 @@ thing-type.chatgpt.account.description = Account at OpenAI that is used for acce
+thing-type.config.chatgpt.account.apiUrl.option.https\://api.openai.com/v1/chat/completions = https://api.openai.com/v1/chat/completions
+thing-type.config.chatgpt.account.modelUrl.label = Model URL
+thing-type.config.chatgpt.account.modelUrl.description = The model url where to retrieve the available models from.
+thing-type.config.chatgpt.account.modelUrl.option.https\://api.openai.com/v1/models = https://api.openai.com/v1/models
```
```suggestion
```
